spec.md

```diff
@@ -2961,6 +2961,104 @@ For this reason, it is an error to use `rwprobe`{.firrtl} on any port on a publi
 
 Probes may not be inputs to modules.
 
+# Contracts
+
+The `contract`{.firrtl} declaration provides a mechanism to verify a predicate or property, and to then use this predicate or property to simplify the circuit for further verification or simulation.
+
+``` firrtl,notest
+FIRRTL version 4.2.0
+circuit Foo:
+  public module Foo:
+    input a: UInt<42>
+    input b: UInt<1337>
+    output x: UInt<42>
+    output y: UInt<1337>
+    ;; snippetbegin
+    contract c, d = a, b:
+      ; contract body
+    connect x, c
+    connect y, d
+    ;; snippetend
+```
+
+A contract conceptually acts like a `node`{.firrtl}, defining one or more names and assigning a value to each.
+It also has a block to hold statements that describe the predicates and properties to be verified.
+Most commonly the `circt_verif_require` and `circt_verif_ensure` intrinsics ([@sec:intrinsics]) are used to describe pre- and post-conditions of the contract.
+These intrinsics can only be used in the body block of a contract.
+The types of the defined names are the types of the expressions given in the definition and must be a ground type or a potentially nested aggregate type of only ground types.
+Probes, properties, and bundles with flips cannot be used in a contract.
```

Comment:
Can the PR reorder this discussion slightly. Specifically, I wouldn't lead with the example, but end with the example. See some of the structure for how layers or layer blocks are discussed.

Roughly:

1. FIRRTL has contracts.
2. This is what a contract is.
3. This is how you declare a contract.
4. This is how you associate node assignments to a contract.
5. These are the restrictions around a contract.
6. This is an example.